File: shell/services/bunny_shell/bunny_files.py
# SPDX-FileCopyrightText: 2026 ComradeArt
# SPDX-License-Identifier: GPL-3.0-or-later
"""Bunny Files: companion-aware file flows. Trust for opens. Not a chatbot.

Host twin of ``lib/bunnyFiles.js``. Nautilus remains the file browser.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping, Sequence
from urllib.parse import unquote

from .command_surface import route_command_answer
from .trust_copy import (
    ALLOW_ONCE_LABEL,
    DONT_ALLOW_LABEL,
    FILE_NOT_UPLOADED,
    FILE_OPEN_BUBBLE,
    FILE_OPEN_DENIED,
    FILE_OPEN_DRIFT,
    FILE_OPEN_EXPIRED,
    FILE_OPEN_FAILED,
    FILE_OPEN_GRANTED,
    FILE_OPEN_HEADLINE,
    NETWORK_ALLOWLIST_NOTE,
)

FILE_ACTIONS = ("ask", "summarise", "workspace", "provenance", "checkpoint", "open")
FILE_ACTION_LABELS = {
    "ask": "Ask Bunny about this file",
    "summarise": "Summarise with Bunny",
    "workspace": "Open in Bunny workspace",
    "provenance": "Show provenance",
    "checkpoint": "Create checkpoint before changes",
    "open": "Open",
}
_BUBBLE_ACTIONS = frozenset({"ask", "provenance"})
_TASK_ACTIONS = frozenset({"summarise", "workspace", "checkpoint"})


def _file_name(path: str) -> str:
    raw = str(path or "").replace("file://", "").strip()
    if not raw:
        return ""
    parts = [part for part in raw.split("/") if part]
    return parts[-1] if parts else raw


def parse_files_uri(uri: str) -> dict[str, Any]:
    text = str(uri or "").strip()
    prefix = "bunny://files/"
    if not text.lower().startswith(prefix):
        return {"valid": False, "action": "", "paths": [], "companionRequired": False}
    rest = text[len(prefix) :]
    action, _, query = rest.partition("?")
    action = action.strip().casefold()
    selection = ""
    if query.startswith("selection="):
        selection = query[len("selection=") :]
    try:
        decoded = unquote(selection)
    except Exception:  # noqa: BLE001 - malformed escape stays empty
        decoded = selection
    paths = [item.strip() for item in decoded.split("\n") if item.strip()]
    valid = action in FILE_ACTIONS
    return {
        "valid": valid,
        "action": action if valid else "",
        "paths": paths,
        "companionRequired": False,
        "chatbot": False,
    }


def file_open_needs_trust(*, action: str = "open", approved_location: bool = False) -> bool:
    if action == "open":
        return True
    if action == "checkpoint":
        return True
    return (not approved_location) and action == "workspace"


def granted_file_path(path: str) -> str:
    """The exact file Trust granted. Not a folder walk, not Pictures."""
    raw = str(path or "").strip()
    if not raw or "\n" in raw or "\0" in raw:
        return ""
    if any(token in raw for token in (";", "|", "`", "$(")):
        return ""
    if raw.casefold().startswith("file://"):
        rest = raw[7:]
        try:
            rest = unquote(rest)
        except Exception:  # noqa: BLE001 - malformed escape is not a grant
            return ""
        if not rest:
            return ""
        raw = rest if rest.startswith("/") else f"/{rest}"
    if not raw.startswith("/"):
        return ""
    return raw


_ONCE_GRANTS = frozenset({"allow", "allow-once", "allow once", "once", "granted"})
_DENY_VERDICTS = frozenset({
    "deny", "denied", "don't allow", "dont-allow", "dont allow",
    "close", "closed", "escape", "dismiss", "cancel",
})
_EXPIRED_VERDICTS = frozenset({"timeout", "expired"})
_BROADEN_VERDICTS = frozenset({
    "always", "session", "allow-session", "allow while using", "allow-while-using",
})
_DEFAULT_APP = frozenset({"", "an application", "application", "default"})


def _is_once_grant(decision: str) -> bool:
    return str(decision or "").strip().casefold() in _ONCE_GRANTS


def _normalize_application(application: str) -> str:
    token = str(application or "").strip()
    if token.casefold() in _DEFAULT_APP:
        return ""
    return token


def resolve_file_open_after_trust(
    *,
    decision: str = "deny",
    path: str = "",
    approved_path: str = "",
    application: str = "",
    approved_application: str = "",
    action: str = "open",
    extra_paths: Sequence[str] | None = None,
) -> dict[str, Any]:
    """Plan the open. Deny-by-default. One file + app. Nothing uploaded."""
    del extra_paths  # never widened — Pictures and siblings stay closed
    offered = granted_file_path(path)
    approved = granted_file_path(approved_path or path)
    offered_app = _normalize_application(application)
    approved_app = _normalize_application(approved_application or application)
    token = str(decision or "").strip().casefold()
    once = token in _ONCE_GRANTS
    closed = {
        "shouldOpen": False,
        "uploaded": False,
        "path": "",
        "application": "",
        "command": None,
        "companionRequired": False,
        "chatbot": False,
        "note": FILE_NOT_UPLOADED,
    }
    if str(action or "open") != "open":
        return {
            **closed,
            "message": FILE_OPEN_DENIED if not once else "Allow once is recorded for this request.",
        }
    if token in _EXPIRED_VERDICTS:
        return {**closed, "message": FILE_OPEN_EXPIRED}
    if token in _DENY_VERDICTS or token in _BROADEN_VERDICTS or not once:
        return {**closed, "message": FILE_OPEN_DENIED}
    if not offered or not approved or offered != approved or offered_app != approved_app:
        return {**closed, "message": FILE_OPEN_DRIFT if (approved or approved_app) else FILE_OPEN_FAILED}
    return {
        "shouldOpen": True,
        "uploaded": False,
        "path": approved,
        "application": approved_app,
        "command": ["gio", "open", approved],
        "companionRequired": False,
        "chatbot": False,
        "note": FILE_NOT_UPLOADED,
        "message": FILE_OPEN_GRANTED,
    }


def apply_file_open_after_trust(
    *,
    decision: str = "deny",
    path: str = "",
    approved_path: str = "",
    application: str = "",
    approved_application: str = "",
    action: str = "open",
    extra_paths: Sequence[str] | None = None,
    opener: Any = None,
) -> dict[str, Any]:
    """Grant opens that file + app. Deny / close / timeout open nothing."""
    plan = resolve_file_open_after_trust(
        decision=decision,
        path=path,
        approved_path=approved_path,
        application=application,
        approved_application=approved_application,
        action=action,
        extra_paths=extra_paths,
    )
    if not plan["shouldOpen"]:
        return {**plan, "launched": False, "paths": []}
    launched = False
    if callable(opener):
        try:
            launched = opener(plan["command"], plan["path"], plan["application"]) is True
        except TypeError:
            try:
                launched = opener(plan["command"], plan["path"]) is True
            except Exception:  # noqa: BLE001 - a failed handler is a failed open
                launched = False
        except Exception:  # noqa: BLE001 - a failed handler is a failed open
            launched = False
    return {
        **plan,
        "launched": launched,
        "paths": [plan["path"]] if launched else [],
        "message": FILE_OPEN_GRANTED if launched else FILE_OPEN_FAILED,
    }


def build_file_open_trust(
    *,
    path: str = "",
    application: str = "an application",
    request_id: str = "file-open",
    cloud_context: str = "none",
    offering_remote_dispatch: bool = False,
    network: str = "Off",
) -> dict[str, Any]:
    name = _file_name(path) or "this file"
    heading = f"Bunny wants to open {name} in {application}" if application else FILE_OPEN_HEADLINE
    return {
        "kind": "TrustPrompt",
        "source": "files",
        "requestId": request_id,
        "heading": heading,
        "category": "files",
        "resource": str(path or name),
        "application": application,
        "focusSafeAnswer": True,
        "allowLabel": ALLOW_ONCE_LABEL,
        "denyLabel": DONT_ALLOW_LABEL,
        "bubbleText": FILE_OPEN_BUBBLE,
        "note": FILE_NOT_UPLOADED,
        "network": network,
        "networkHonesty": NETWORK_ALLOWLIST_NOTE,
        "cloudContext": cloud_context,
        "offeringRemoteDispatch": bool(offering_remote_dispatch),
        "chatbot": False,
        "transcript": False,
        "companionRequired": False,
    }


def _caption(action: str, paths: Sequence[str]) -> str:
    count = len(paths)
    first = _file_name(paths[0]) if paths else "this file"
    if action == "ask":
        if count > 1:
            return f"Ask Bunny about {count} files. Nothing is uploaded."
        return f"Ask Bunny about {first}. Nothing is uploaded."
    if action == "summarise":
        return f"Summarise {first} here. Longer notes open a task card."
    if action == "workspace":
        return f"Attach {first} to a Bunny workspace. Files stay on this computer."
    if action == "provenance":
        return f"Provenance for {first}."
    if action == "checkpoint":
        return f"Checkpoint before changing {first}."
    return f"Open {first}. Review the request."


@dataclass(frozen=True)
class FilesRoute:
    action: str
    paths: tuple[str, ...]
    surface: str
    needs_trust: bool
    bubble_text: str
    transcript: bool = False
    chatbot: bool = False
    companion_required: bool = False
    uploaded: bool = False


def route_files_action(
    *,
    action: str = "ask",
    paths: Sequence[str] | None = None,
    application: str = "",
    approved_location: bool = True,
    cloud_context: str = "none",
    offering_remote_dispatch: bool = False,
) -> dict[str, Any]:
    resolved = action if action in FILE_ACTIONS else "ask"
    listed = tuple(str(item) for item in (paths or ()) if str(item).strip())
    needs_trust = file_open_needs_trust(action=resolved, approved_location=approved_location)
    caption = _caption(resolved, listed)
    routed = route_command_answer(caption, working=resolved in _TASK_ACTIONS, title=FILE_ACTION_LABELS[resolved])
    surface = "trust" if needs_trust else ("bubble" if resolved in _BUBBLE_ACTIONS else routed.surface)
    return {
        "action": resolved,
        "paths": list(listed),
        "application": application,
        "surface": surface,
        "needsTrust": needs_trust,
        "bubbleText": FILE_OPEN_BUBBLE if needs_trust else caption,
        "trust": build_file_open_trust(
            path=listed[0] if listed else "",
            application=application,
            cloud_context=cloud_context,
            offering_remote_dispatch=offering_remote_dispatch,
        )
        if needs_trust
        else None,
        "transcript": False,
        "chatbot": False,
        "companionRequired": False,
        "uploaded": False,
        "note": FILE_NOT_UPLOADED,
    }


def build_bunny_files(
    *,
    locations: Sequence[Mapping[str, Any] | str] | None = None,
    entries: Sequence[Mapping[str, Any] | str] | None = None,
    query: str = "",
    reduced_motion: bool = False,
) -> dict[str, Any]:
    rows = []
    for index, entry in enumerate(list(entries or ())[:40]):
        item = dict(entry) if isinstance(entry, Mapping) else {"path": entry}
        path = str(item.get("path") or item.get("uri") or "")
        rows.append({
            "id": f"file-{index}",
            "title": str(item.get("name") or _file_name(path) or path),
            "subtitle": path,
            "kind": str(item.get("kind") or "file"),
            "path": path,
            "canFocus": True,
        })
    places = []
    for item in locations or ():
        rec = dict(item) if isinstance(item, Mapping) else {"path": item}
        places.append({"path": str(rec.get("path") or ""), "enabled": rec.get("enabled", True) is not False})
    return {
        "kind": "BunnyFiles",
        "title": "Bunny Files",
        "summary": "Approved folders and companion actions. Nautilus still browses. Opens need Trust.",
        "query": str(query or ""),
        "locations": places,
        "rows": rows,
        "actions": list(FILE_ACTIONS),
        "companionRequired": False,
        "chatbot": False,
        "transcript": False,
        "nautilusIsBrowser": True,
        "uploaded": False,
        "note": FILE_NOT_UPLOADED,
        "networkHonesty": NETWORK_ALLOWLIST_NOTE,
        "reducedMotion": bool(reduced_motion),
        "motionMs": 0 if reduced_motion else 220,
        "accessibleName": "Bunny Files",
        "canFocus": True,
    }

File: shell/components/gnome-shell-extension/lib/bunnyFiles.js
// SPDX-FileCopyrightText: 2026 ComradeArt
// SPDX-License-Identifier: GPL-3.0-or-later
//
// Bunny Files: companion-aware file flows. Trust for opens. Bubbles, not a
// chatbot wall. This is the host-testable twin of bunny_files.py.
//
// Nautilus remains the file browser. This module is the Bunny overlay:
// Ask / Summarise / Workspace / Provenance / Checkpoint / Open. Open is a
// Trust question (Allow once / Don't allow, Don't allow focused). Nothing
// here uploads a file.

import {
    buildBubble, buildList, buildSearchField, buildSheet, limitBubbleText,
} from './design/primitives.js';
import {SCREEN_QUESTIONS} from './design/tokens.js';
import {
    ALLOW_ONCE_LABEL, DONT_ALLOW_LABEL, FILE_NOT_UPLOADED, FILE_OPEN_BUBBLE,
    FILE_OPEN_DENIED, FILE_OPEN_FAILED, FILE_OPEN_GRANTED, FILE_OPEN_HEADLINE,
    NETWORK_ALLOWLIST_NOTE, buildPrompt,
} from './trustPrompt.js';
import {routeCommandAnswer} from './commandSurface.js';

export {
    FILE_NOT_UPLOADED, FILE_OPEN_BUBBLE, FILE_OPEN_DENIED, FILE_OPEN_FAILED,
    FILE_OPEN_GRANTED, FILE_OPEN_HEADLINE,
};

export const FILE_ACTIONS = Object.freeze([
    'ask', 'summarise', 'workspace', 'provenance', 'checkpoint', 'open',
]);

export const FILE_ACTION_LABELS = Object.freeze({
    ask: 'Ask Bunny about this file',
    summarise: 'Summarise with Bunny',
    workspace: 'Open in Bunny workspace',
    provenance: 'Show provenance',
    checkpoint: 'Create checkpoint before changes',
    open: 'Open',
});

const BUBBLE_ACTIONS = new Set(['ask', 'provenance']);
const TASK_ACTIONS = new Set(['summarise', 'workspace', 'checkpoint']);

function fileName(path) {
    const raw = String(path || '').replace(/^file:\/\//, '').trim();
    if (!raw)
        return '';
    const parts = raw.split('/').filter(Boolean);
    return parts[parts.length - 1] || raw;
}

function decodeSelection(value) {
    const raw = String(value ?? '');
    if (!raw)
        return [];
    try {
        return decodeURIComponent(raw)
            .split('\n')
            .map(item => item.trim())
            .filter(Boolean);
    } catch {
        return raw.split('\n').map(item => item.trim()).filter(Boolean);
    }
}

/**
 * Parse a Nautilus handoff `bunny://files/{action}?selection=...`.
 */
export function parseFilesUri(uri) {
    const text = String(uri || '').trim();
    const match = /^bunny:\/\/files\/([a-z]+)(?:\?selection=([^#]*))?$/i.exec(text);
    if (!match)
        return {valid: false, action: '', paths: [], companionRequired: false};
    const action = match[1].toLowerCase();
    return {
        valid: FILE_ACTIONS.includes(action),
        action: FILE_ACTIONS.includes(action) ? action : '',
        paths: decodeSelection(match[2] || ''),
        companionRequired: false,
        chatbot: false,
    };
}

/**
 * Opening a file in an application is always a Trust question.
 *
 * Browse of an already-approved location is not. Ask/summarise stay local
 * unless the caller is offering a remote hop — then the existing
 * remote_dispatch disclosure applies.
 */
export function fileOpenNeedsTrust({action = 'open', approvedLocation = false} = {}) {
    if (String(action) === 'open')
        return true;
    if (String(action) === 'checkpoint')
        return true;
    return !approvedLocation && String(action) === 'workspace';
}

/**
 * The exact file Trust granted. Not a folder walk, not Pictures.
 */
export function grantedFilePath(path) {
    let raw = String(path || '').trim();
    if (!raw || raw.includes('\n') || raw.includes('\0'))
        return '';
    if (/[;|`]|\$\(/.test(raw))
        return '';
    if (raw.toLowerCase().startsWith('file://')) {
        let rest = raw.slice(7);
        try {
            rest = decodeURIComponent(rest);
        } catch {
            return '';
        }
        if (!rest)
            return '';
        raw = rest.startsWith('/') ? rest : `/${rest}`;
    }
    if (!raw.startsWith('/'))
        return '';
    return raw;
}

function isOnceGrant(decision) {
    const token = String(decision || '').trim().toLowerCase();
    return token === 'allow' || token === 'allow-once' || token === 'allow once'
        || token === 'once' || token === 'granted';
}

/**
 * Plan the open. Deny-by-default. One path. Nothing uploaded.
 */
export function resolveFileOpenAfterTrust({
    decision = 'deny', path = '', action = 'open', extraPaths = [],
} = {}) {
    void extraPaths; // never widened — Pictures and siblings stay closed
    const granted = grantedFilePath(path);
    const once = isOnceGrant(decision);
    if (String(action || 'open') !== 'open') {
        return {
            shouldOpen: false,
            uploaded: false,
            path: '',
            command: null,
            companionRequired: false,
            chatbot: false,
            note: FILE_NOT_UPLOADED,
            message: once ? 'Allow once is recorded for this request.' : FILE_OPEN_DENIED,
        };
    }
    if (!once || !granted) {
        return {
            shouldOpen: false,
            uploaded: false,
            path: '',
            command: null,
            companionRequired: false,
            chatbot: false,
            note: FILE_NOT_UPLOADED,
            message: !once ? FILE_OPEN_DENIED : FILE_OPEN_FAILED,
        };
    }
    return {
        shouldOpen: true,
        uploaded: false,
        path: granted,
        command: ['gio', 'open', granted],
        companionRequired: false,
        chatbot: false,
        note: FILE_NOT_UPLOADED,
        message: FILE_OPEN_GRANTED,
    };
}

/**
 * Grant opens that file. Deny opens nothing. Host-testable via `opener`.
 */
export function applyFileOpenAfterTrust(record = {}, opener = null) {
    const plan = resolveFileOpenAfterTrust(record);
    if (!plan.shouldOpen)
        return {...plan, launched: false, paths: []};
    let launched = false;
    if (typeof opener === 'function') {
        try {
            launched = opener(plan.command, plan.path) === true;
        } catch {
            launched = false;
        }
    }
    return {
        ...plan,
        launched,
        paths: launched ? [plan.path] : [],
        message: launched ? FILE_OPEN_GRANTED : FILE_OPEN_FAILED,
    };
}

export function buildFileOpenTrust({
    path = '', application = 'an application', requestId = 'file-open',
    cloudContext = 'none', offeringRemoteDispatch = false, network = 'Off',
} = {}) {
    const name = fileName(path) || 'this file';
    const heading = application
        ? `Bunny wants to open ${name} in ${application}`
        : FILE_OPEN_HEADLINE;
    const record = {
        requestId,
        headline: heading,
        category: 'files',
        categoryTitle: 'Files',
        resource: String(path || name),
        capabilityNote: 'Open this file in that application for this request only.',
        reason: FILE_NOT_UPLOADED,
        options: [{scope: 'once', label: ALLOW_ONCE_LABEL}],
        denyOption: {label: DONT_ALLOW_LABEL},
        risk: 'medium',
        cloudContext,
        offeringRemoteDispatch,
        fileAccess: String(path || name),
        network,
    };
    const prompt = buildPrompt(record);
    prompt.bubble = buildBubble({
        text: FILE_OPEN_BUBBLE,
        actions: [
            {id: 'review', label: 'Review the request'},
        ],
    });
    prompt.chatbot = false;
    prompt.transcript = false;
    prompt.companionRequired = false;
    prompt.focusSafeAnswer = true;
    return prompt;
}

function captionFor(action, paths) {
    const count = paths.length;
    const first = fileName(paths[0]) || 'this file';
    if (action === 'ask')
        return count > 1
            ? `Ask Bunny about ${count} files. Nothing is uploaded.`
            : `Ask Bunny about ${first}. Nothing is uploaded.`;
    if (action === 'summarise')
        return `Summarise ${first} here. Longer notes open a task card.`;
    if (action === 'workspace')
        return `Attach ${first} to a Bunny workspace. Files stay on this computer.`;
    if (action === 'provenance')
        return `Provenance for ${first}.`;
    if (action === 'checkpoint')
        return `Checkpoint before changing ${first}.`;
    return `Open ${first}. Review the request.`;
}

/**
 * Route one Files action. Short copy stays in the bubble. Opens go through
 * Trust. This is not a chat transcript.
 */
export function routeFilesAction({
    action = 'ask', paths = [], application = '', approvedLocation = true,
    cloudContext = 'none', offeringRemoteDispatch = false,
} = {}) {
    const resolved = FILE_ACTIONS.includes(action) ? action : 'ask';
    const list = Array.isArray(paths) ? paths.map(item => String(item)).filter(Boolean) : [];
    const needsTrust = fileOpenNeedsTrust({action: resolved, approvedLocation});
    const caption = limitBubbleText(captionFor(resolved, list));
    const routed = routeCommandAnswer(caption, {
        working: TASK_ACTIONS.has(resolved),
        title: FILE_ACTION_LABELS[resolved],
    });
    const surface = needsTrust ? 'trust' : (BUBBLE_ACTIONS.has(resolved) ? 'bubble' : routed.surface);
    return {
        action: resolved,
        paths: list,
        surface,
        needsTrust,
        bubble: buildBubble({text: needsTrust ? FILE_OPEN_BUBBLE : caption}),
        trust: needsTrust
            ? buildFileOpenTrust({
                path: list[0] || '',
                application,
                cloudContext,
                offeringRemoteDispatch,
            })
            : null,
        taskCard: surface === 'task-card' ? routed.taskCard : null,
        transcript: false,
        chatbot: false,
        companionRequired: false,
        uploaded: false,
        note: FILE_NOT_UPLOADED,
    };
}

/**
 * The Bunny Files overlay: a sheet of approved locations, a search field,
 * and companion actions. Not a chatbot. Companion is optional.
 */
export function buildBunnyFiles({
    locations = [], entries = [], query = '', reducedMotion = false,
} = {}) {
    const rows = (Array.isArray(entries) ? entries : []).slice(0, 40).map((entry, index) => {
        const item = entry && typeof entry === 'object' ? entry : {path: entry};
        const path = String(item.path || item.uri || '');
        return {
            id: `file-${index}`,
            title: item.name || fileName(path) || path,
            subtitle: path,
            kind: item.kind || 'file',
            path,
            accessibleName: `${item.name || fileName(path) || 'File'}. ${item.kind || 'file'}`,
            canFocus: true,
        };
    });
    const places = (Array.isArray(locations) ? locations : []).map(item => {
        const rec = item && typeof item === 'object' ? item : {path: item};
        return {
            path: String(rec.path || ''),
            enabled: rec.enabled !== false,
        };
    });
    return {
        kind: 'BunnyFiles',
        title: 'Bunny Files',
        summary: 'Approved folders and companion actions. Nautilus still browses. Opens need Trust.',
        field: buildSearchField({
            value: query,
            placeholder: 'Find a file in approved folders',
        }),
        sheet: buildSheet({
            title: 'Files',
            actions: FILE_ACTIONS.map(id => ({
                id,
                label: FILE_ACTION_LABELS[id],
                variant: id === 'open' ? 'primary' : 'secondary',
                accessibleName: FILE_ACTION_LABELS[id],
            })),
        }),
        list: buildList({rows}),
        locations: places,
        actions: FILE_ACTIONS.slice(),
        questions: SCREEN_QUESTIONS,
        doing: 'Browsing approved files',
        bunny: 'Ready to ask or open with Trust',
        next: 'Select a file. Open asks Allow once or Don\'t allow.',
        companionRequired: false,
        chatbot: false,
        transcript: false,
        nautilusIsBrowser: true,
        uploaded: false,
        note: FILE_NOT_UPLOADED,
        networkHonesty: NETWORK_ALLOWLIST_NOTE,
        reducedMotion: Boolean(reducedMotion),
        motionMs: reducedMotion ? 0 : 220,
        accessibleName: 'Bunny Files',
        canFocus: true,
        styleClass: 'bunny-files bunny-sheet',
    };
}
